src/engine/sparqlExpressions/StringExpressions.cpp

```diff
@@ -401,22 +405,74 @@ class ConcatExpression : public detail::VariadicExpression {
     // We store the (intermediate) result either as single string or a vector.
     // If the result is a string, then all the previously evaluated children
     // were constants (see above).
+
     std::variant<std::string, StringVec> result{std::string{""}};
+    std::optional<std::string> datatypeOrLangTag;
+
+    // Extracts the datatype or language tag from a literal.
+    // Language tags are returned as "@lang",datatypes as IRIs without angle
+    // brackets.
+    auto getDatatypeOrLanguageTag = [&](const auto& literal) {
+      if (literal.hasDatatype()) {
+        return std::string{asStringViewUnsafe(literal.getDatatype())};
+      } else if (literal.hasLanguageTag()) {
+        return "@" + std::string{asStringViewUnsafe(literal.getLanguageTag())};
+      }
+      return std::string{};
+    };
+
+    auto getCommonDescriptor = [&](const std::optional<std::string>& str1,
+                                   const std::string& str2) {
+      if (!str1.has_value() ||
+          (!str1.value().empty() && str1.value() == str2)) {
+        return str2;
+      }
+      return std::string{};
+    };
+
+    // Builds a literal from the given string content and descriptor.
+    // If descriptor starts with "@", it's a language tag. Otherwise it's either
+    // the datatype std:string or empty.
+    auto toLiteralFromDescriptor =
+        [](const std::string& content,
+           const std::optional<std::string>& descriptor) {
+          if (descriptor.has_value() && descriptor.value().starts_with("@")) {
+            return toLiteral(content, descriptor);
+          } else if (descriptor.has_value() && !descriptor.value().empty()) {
+            return toLiteral(
+                content, Iri::fromIrirefWithoutBrackets(descriptor.value()));
+          }
+          return toLiteral(content);
+        };
```

Comment:
I think the most efficient way, and also the 
easiest for code quality is to simply add a `concat` function to the `Literal` class, that implements this, so you can simply write `literalSoFar.concat(nextLiteral)`.
This function can then do all the adjustments of the datatypes and language tags, but never has to do unnecessary allocations (the one for concating strings is somewhat unavoidable).

Reply:
With this change, the code in this evaluate function is again really simple and similar to the original one.